src/btree/bt_cursor.c

```diff
@@ -232,107 +232,197 @@ __cursor_fix_implicit(WT_BTREE *btree, WT_CURSOR_BTREE *cbt)
+        key = cbt->tmp;
 
     /*
-     * We may be pointing to an insert object, and we may have a page with
-     * existing entries.  Insert objects always have associated update
-     * objects (the value).  Any update object may be deleted, or invisible
-     * to us.  In the case of an on-page entry, there is by definition a
-     * value that is visible to us, the original page cell.
-     *
-     * If we find a visible update structure, return our caller a reference
-     * to it because we don't want to repeatedly search for the update, it
-     * might suddenly become invisible (imagine a read-uncommitted session
-     * with another session's aborted insert), and we don't want to handle
-     * that potential error every time we look at the value.
-     *
-     * Unfortunately, the objects we might have and their relationships are
-     * different for the underlying page types.
+     * In the case of row-store, an insert object implies ignoring any page objects, no insert
+     * object can have the same key as an on-page object.
      *
-     * In the case of row-store, an insert object implies ignoring any page
-     * objects, no insert object can have the same key as an on-page object.
      * For row-store:
      *	if there's an insert object:
      *		if there's a visible update:
      *			exact match
      *		else
      *			no exact match
      *	else
-     *		use the on-page object (which may have an associated
-     *		update object that may or may not be visible to us).
+     *		use the on-page object (which may have an associated update object that may or may
+     *		not be visible to us).
+     */
+    WT_RET(__cursor_valid_insert(cbt, key, valid, check_bounds));
+    if (*valid || cbt->ins != NULL)
+        return (0);
+
+    /* The search function doesn't check for empty pages. */
+    if (page->entries == 0)
+        return (0);
+    /*
+     * In case of prepare conflict, the slot might not have a valid value, if the update in the
+     * insert list of a new page scanned is in prepared state.
+     */
+    WT_ASSERT(session, cbt->slot == UINT32_MAX || cbt->slot < page->entries);
+
+    /*
+     * The key can be NULL only when we didn't find an exact match, copy the search found key into
+     * the temporary buffer for further use.
+     */
+    if (key == NULL) {
+        WT_RET(__wt_row_leaf_key(session, page, &page->pg_row[cbt->slot], cbt->tmp, true));
+        key = cbt->tmp;
+    }
+    if (check_bounds) {
+        WT_RET(__btcur_bounds_contains_key(
+          session, &cbt->iface, key, WT_RECNO_OOB, &key_out_of_bounds, NULL));
+        /* The key value pair we were trying to return weren't within the given bounds. */
+        if (key_out_of_bounds)
+            return (0);
+    }
+
+    /* Check for an update. */
+    upd = (page->modify != NULL && page->modify->mod_row_update != NULL) ?
+      page->modify->mod_row_update[cbt->slot] :
+      NULL;
+
+    WT_RET(__cursor_valid_ondisk(cbt, key, upd, valid));
```

Comment:
Should we just do `return (__cursor_valid_ondisk(cbt, key, upd, valid)`?

Reply:
f0703bca8